tools/xmlschema.py

```diff
@@ -0,0 +1,312 @@
+    "unsigned int": "unsignedInt",
+    "unsigned long": "unsignedLong",
+}
+
+# Mapping between "required" values found in SDF files to the corresponding
+# minOccurs and maxOccurs found in XSD
+SDF_REQUIRED_TO_MIN_MAX_OCCURS: Dict[str, Tuple[str, str]] = {
+    "0": ("0", "1"),  # Required: 0,  (minOccurs: 0, maxOccurs: 1)
+    "1": ("1", "1"),  # Required: 1,  (minOccurs: 1, maxOccurs: 1)
+    "+": ("1", "unbounded"),  # Required: +,  (minOccurs: 1, maxOccurs: inf)
+    "*": ("0", "unbounded"),  # Required: *,  (minOccurs: 0, maxOccurs: inf)
+    "-1": ("0", "unbounded"),  # Required: -1, (minOccurs: 0, maxOccurs: inf)
+}
+
+
+def indent_lines(lines: List[str], indent: int) -> List[str]:
+    """
+    Indent a list of xml lines group of lines by a number (indent) of spaces
+
+    """
+    return [" " * indent + line for line in lines]
+
+
+def get_attribute(element: ElementTree.Element, attrib: str) -> Optional[str]:
+    """
+    Retrieve XML attribute from an element
+    """
+    return element.attrib[attrib] if attrib in element.attrib else None
+
+
+def is_std_type(sdf_type: str) -> bool:
+    """
+    Check if sdf_type is a known XSD standard type.
+    Return true if the sdf_type is in the set of known types, false otherwise.
+    """
+    return sdf_type in SDF_TYPES_TO_XSD_STD_TYPES
+
+
+def xsd_type_string(sdf_type: str) -> Optional[str]:
+    """
+    Check if xsd_type is a known XSD standard type.
+    If it is, return 'xsd:' + type, None otherwise.
+    """
+    if is_std_type(sdf_type):
+        xsd_type = SDF_TYPES_TO_XSD_STD_TYPES[sdf_type]
+        return "xsd:" + xsd_type
+    return None
+
+
+def print_documentation(element: ElementTree.Element) -> List[str]:
+    """
+    Print the documentation associated with an element
+    """
+    lines = []
+    description = element.find("description")
+    if (
+        description is not None
+        and description.text is not None
+        and len(description.text)
+    ):
+        lines.append("<xsd:annotation>")
+        lines.append("  <xsd:documentation xml:lang='en'>")
+        lines.append(f"    <![CDATA[{description.text}]]>")
+        lines.append("  </xsd:documentation>")
+        lines.append("</xsd:annotation>")
+    return lines
+
+
+def print_include(element: ElementTree.Element) -> List[str]:
+    """
+    Print include tag information
+    """
+    lines = []
+    filename = get_attribute(element, "filename")
+    if filename is not None:
+        loc = "http://sdformat.org/schemas/"
+        loc += filename.replace(".sdf", ".xsd")
+        lines.append(f"<xsd:include schemaLocation='{loc}'/>")
+    return lines
+
+
+def print_include_ref(element: ElementTree.Element, sdf_root_dir: str) -> List[str]:
+    """
+    Print include tag reference information
+    """
+    lines = []
+    filename = get_attribute(element, "filename")
+    if filename is not None:
+        sdf_path = os.path.join(sdf_root_dir, filename)
+        if os.path.exists(sdf_path):
```

Comment:
The ruby code would have thrown an exception or something if the file wasn't 
found. We're silently ignoring it here.